webhookURLValidator.go

```diff
@@ -0,0 +1,205 @@
+	errInvalidSubnet         = errors.New("invalid subnet")
+)
+
+// filterNil takes out all entries of Nil value from the slice.
+func filterNil(vs []ValidURLFunc) (filtered []ValidURLFunc) {
+	for _, v := range vs {
+		if v != nil {
+			filtered = append(filtered, v)
+		}
+	}
+	return
+}
+
+// GoodConfigURL parses the given webhook's Config.URL
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodConfigURL(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		parsedURL, err := url.ParseRequestURI(w.Config.URL)
+		if err != nil {
+			return fmt.Errorf("%w: %v", errInvalidURL, err)
+		}
+		for _, f := range vs {
+			err = f(parsedURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidURL, err)
+			}
+		}
+		return nil
+	}
+}
+
+// GoodFailureURL parses the given webhook's FailureURL
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodFailureURL(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		if w.FailureURL != "" {
+			parsedFailureURL, err := url.ParseRequestURI(w.FailureURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+			}
+			for _, f := range vs {
+				err = f(parsedFailureURL)
+				if err != nil {
+					return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+				}
+			}
+		}
+		return nil
+	}
+}
+
+// GoodAlternativeURLs parses the given webhook's Config.AlternativeURLs
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodAlternativeURLs(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		for _, u := range w.Config.AlternativeURLs {
+			if u == "" {
+				return errInvalidAlternativeURL
+			}
+			parsedAlternativeURL, err := url.ParseRequestURI(u)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+			}
+			for _, f := range vs {
+				err = f(parsedAlternativeURL)
+				if err != nil {
+					return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+				}
+			}
+		}
+		return nil
+	}
+}
+
+// HTTPSOnlyEndpoints creates a ValidURLFunc that considers a URL valid if the scheme
+// of the address is https.
+func HTTPSOnlyEndpoints() ValidURLFunc {
+	return func(u *url.URL) error {
+		if u.Scheme != "https" {
+			return errURLIsNotHTTPS
+		}
+		return nil
+	}
+}
+
+// RejectHosts creates a ValidURLFunc that checks the URL and ensures the
+// host does not contain any strings in the list of invalid hosts. It returns an error
+// if the host does include an invalid host name.
+func RejectHosts(invalidHosts []string) ValidURLFunc {
+	ih := []string{}
+	for _, v := range invalidHosts {
+		if v != "" {
+			ih = append(ih, v)
+		}
+	}
+	return func(u *url.URL) error {
+		host := u.Host
+		for _, v := range ih {
+			if strings.Contains(host, v) {
```

Comment:
since `Host` is just a field (and not some possibly expensive getter method), I think it's ok to just call Contains with `u.Host` instead of creating a local var.